File: crates/zeroclaw-plugin-sdk/src/messaging.rs
//! Messaging module — wraps send_message and get_channels host functions.
//!
//! Each function serializes a typed request to JSON, calls the corresponding
//! host function via Extism shared memory, and deserializes the JSON response.

use extism_pdk::*;
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// Request / response types (mirror the host-side structs)
// ---------------------------------------------------------------------------

#[derive(Debug, Serialize)]
struct SendMessageRequest {
    channel: String,
    recipient: String,
    message: String,
}

#[derive(Debug, Deserialize)]
struct SendMessageResponse {
    success: bool,
    #[serde(default)]
    error: Option<String>,
}

#[derive(Debug, Deserialize)]
struct GetChannelsResponse {
    #[serde(default)]
    channels: Vec<String>,
    #[serde(default)]
    error: Option<String>,
}

// ---------------------------------------------------------------------------
// Host function imports
// ---------------------------------------------------------------------------

#[host_fn]
extern "ExtismHost" {
    fn zeroclaw_send_message(input: Json<SendMessageRequest>) -> Json<SendMessageResponse>;
    fn zeroclaw_get_channels(input: Json<()>) -> Json<GetChannelsResponse>;
}

// ---------------------------------------------------------------------------
// Public wrapper API
// ---------------------------------------------------------------------------

/// Send a message to a recipient on the given channel.
pub fn send(channel: &str, recipient: &str, message: &str) -> Result<(), Error> {
    let request = SendMessageRequest {
        channel: channel.to_string(),
        recipient: recipient.to_string(),
        message: message.to_string(),
    };
    let Json(response) = unsafe { zeroclaw_send_message(Json(request))? };
    if let Some(err) = response.error {
        return Err(Error::msg(err));
    }
    if !response.success {
        return Err(Error::msg("send_message returned success=false"));
    }
    Ok(())
}

/// Get the list of available channel names.
pub fn get_channels() -> Result<Vec<String>, Error> {
    let Json(response) = unsafe { zeroclaw_get_channels(Json(()))? };
    if let Some(err) = response.error {
        return Err(Error::msg(err));
    }
    Ok(response.channels)
}

File: sdks/python/src/zeroclaw_plugin_sdk/messaging.py
"""Messaging module — wraps send_message and get_channels host functions.

Each function serializes a typed request dict to JSON, calls the
corresponding ``zeroclaw_*`` host function via Extism shared memory,
and deserializes the JSON response.
"""

from __future__ import annotations

import json
from typing import List

import extism_pdk as pdk


def send(channel: str, recipient: str, message: str) -> None:
    """Send a message to a recipient on the given channel.

    Raises ``RuntimeError`` if the host reports an error.
    """
    request = json.dumps({
        "channel": channel,
        "recipient": recipient,
        "message": message,
    })
    raw = pdk.host_fn("zeroclaw_send_message", request)
    response = json.loads(raw)
    if response.get("error"):
        raise RuntimeError(response["error"])
    if not response.get("success"):
        raise RuntimeError("send_message returned success=false")


def get_channels() -> List[str]:
    """Get the list of available channel names.

    Returns a Python list of channel name strings.
    Raises ``RuntimeError`` if the host reports an error.
    """
    request = json.dumps({})
    raw = pdk.host_fn("zeroclaw_get_channels", request)
    response = json.loads(raw)
    if response.get("error"):
        raise RuntimeError(response["error"])
    return response.get("channels", [])
